### ISSUE TITLE
twisted.protocols.amp.BinaryBoxProtocol.connectionLost chokes on reason = main.CONNECTION_DONE

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3922]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| dotz reported|
|-|-|
|Trac ID|trac#3922|
|Type|defect|
|Created|2009-07-19 19:12:59Z|
|Branch|https://github.com/twisted/twisted/tree/stdio-connection-lost-reason-3922|

I am running ActivePython 2.6 on Windows Vista with Twisted Trunk 19/07/2009 and ampoule trunk 19/07/2009 with patched https://bugs.launchpad.net/ampoule/+bug/401460

Trying to run ampoule-trunk/examples/pid.py example file shows, that amp.BinaryBoxProtocol chokes when reason is main.CONNECTION_DONE -- amp.py line 1992.

Why does it use reason.check(...) ? Where is that function documented, is it specified by any interface? As far as I know, 'reason' passed to connectionLost may be an ordinary Exception class, so I think there should be a check for that...

```
2009-07-19 21:01:28+0200 [-] FROM 0:    Traceback (most recent call last):
2009-07-19 21:01:28+0200 [-] FROM 0:      File "<string>", line 20, in <module>
2009-07-19 21:01:28+0200 [-] FROM 0:
2009-07-19 21:01:28+0200 [-] FROM 0:      File "<string>", line 19, in main
2009-07-19 21:01:28+0200 [-] FROM 0:
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\internet\base.py", line 1166, in run
2009-07-19 21:01:28+0200 [-] FROM 0:        self.mainLoop()
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\internet\base.py", line 1175, in mainLoop
2009-07-19 21:01:28+0200 [-] FROM 0:        self.runUntilCurrent()
2009-07-19 21:01:28+0200 [-] FROM 0:    --- <exception caught here> ---
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\internet\base.py", line 779, in runUntilCurrent
2009-07-19 21:01:28+0200 [-] FROM 0:        call.func(*call.args, **call.kw)
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\internet\_pollingfile.py", line 72, in _pollEvent
2009-07-19 21:01:28+0200 [-] FROM 0:        workUnits += resource.checkWork()
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\internet\_pollingfile.py", line 129, in checkWork
2009-07-19 21:01:28+0200 [-] FROM 0:        self.cleanup()
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\internet\_pollingfile.py", line 134, in cleanup
2009-07-19 21:01:28+0200 [-] FROM 0:        self.lostCallback()
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\internet\_win32stdio.py", line 59, in readConnectionLost
2009-07-19 21:01:28+0200 [-] FROM 0:        self.checkConnLost()
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\internet\_win32stdio.py", line 73, in checkConnLost
2009-07-19 21:01:28+0200 [-] FROM 0:        self.proto.connectionLost(main.CONNECTION_DONE)
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\ampoule-0.1.1-py2.6.egg\ampoule\child.py", line 12, in connectionLost
2009-07-19 21:01:28+0200 [-] FROM 0:        amp.AMP.connectionLost(self, reason)
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\protocols\amp.py", line 2229, in connectionLost
2009-07-19 21:01:28+0200 [-] FROM 0:        BinaryBoxProtocol.connectionLost(self, reason)
2009-07-19 21:01:28+0200 [-] FROM 0:      File "c:\python26\lib\site-packages\twisted-8.2.0-py2.6-win32.egg\twisted\protocols\amp.py", line 1992, in connectionLost
2009-07-19 21:01:28+0200 [-] FROM 0:        elif reason.check(ConnectionClosed) and self._justStartedTLS:
2009-07-19 21:01:28+0200 [-] FROM 0:    exceptions.AttributeError: 'ConnectionDone' object has no attribute 'check'

```

Attachments:

* [fix-binaryboxprotocol.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3d2/3d2c898fa7e1b44a6391738a0d2a98f197ce98fa/d34bf142085ed14bde60ca2e6e43d3e34e78db92.patch) (819 bytes) - added by dotz on 2009-07-19 19:13:26Z - patch for #3922
* [patch-test-amp-binaryboxprotocol.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3d2/3d2c898fa7e1b44a6391738a0d2a98f197ce98fa/64f9d2d2395413cd21709fa2c9e40b57d18a3b13.patch) (1507 bytes) - added by dotz on 2009-07-19 19:17:34Z - test included
* [fix-3922.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3d2/3d2c898fa7e1b44a6391738a0d2a98f197ce98fa/5fae315ad4ccb8d5f865cd5169df65266d2c3789.patch) (625 bytes) - added by dotz on 2009-11-07 21:19:53Z - New and hopefully proper fix for this issue attached.

<details><summary>Searchable metadata</summary>

```
trac-id__3922 3922
type__defect defect
reporter__dotz dotz
priority__normal normal
milestone__ 
branch__branches_stdio_connection_lost_reason_3922 branches/stdio-connection-lost-reason-3922
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__win32 win32
time__1248030779000000 1248030779000000
changetime__1258920771000000 1258920771000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

The `AMP` code is correct, the bug is in _win32stdio: it should pass a failure to connectionLost, not a exception (ie, failure.Failreu(CONNECTION_DONE), like it's done everywhere else.

**Comment by twisted-trac** [comment]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|dotz commented|
|-|-|

How about we just fix it in the trunk? I don't think we can add tests, that test API-misuse... correct me if I'm wrong (with some details, perhaps) and I'll see if I can come up with a patch.

**Comment by twisted-trac** [comment]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|dotz commented|
|-|-|

Do you need anything else from me to get this one small change merged?

**Comment by twisted-trac** [comment]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|dotz commented|
|-|-|

Do you need anything else from me to get this one small change merged?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to dotz|
|-|-|

The point of tests for API misuse is to test that, when presented with bad input, a library will give helpful debugging error messages and exceptions to the developer who is attempting to use it.  It's just the same as adding tests for good error messages when a user clicks the wrong button, or hits submit without fully filling out a form.

So yes, we can, and do, have plenty of tests for API misuse :).

This ticket doesn't appear to be in review; if you want to get stuff merged, please follow the development process outlined in TwistedDevelopment - i.e. add the 'review' keyword, reassign to nobody (the empty entry in the "assign to" field).

Thanks for all your contributions, sorry that the process has been getting in the way here.  Hopefully we can get all these minor fixes integrated soon.

**Comment by twisted-trac** [comment]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|dotz commented|
|-|-|

New and hopefully proper fix for this issue attached.

I don't have a slightest idea how to create a test for this issue. Really. If we want to do as Glyph said and make the library developer-friendly, we would need a type check for value passed to connectionLost. 

If anybody is able to help me create a test for this thing, please do.

**Comment by twisted-trac** [comment]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dotz removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to dotz|
|-|-|

It seems likely that at least one of the tests in `twisted.test.test_stdio` must cover this already - but is probably missing an assertion about it or is being tricked by the automatic-Failure-wrapping-and-errbacking behavior of `Deferred.callback`.  I'd take a look at the tests there and find one that has its connection lost reason in an easy-to-make-assertions-about place, and then add the assertion there.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [27581]) Change the stdio loseConnection test to also check the failure passed to `connectionLost`

refs [#3922](https://github.com/twisted/twisted/issues/3922)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [27583]) Apply dotz's proposed fix

refs [#3922](https://github.com/twisted/twisted/issues/3922)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Okay, I expanded one of the tests to verify this behavior.  I [built on Windows](http://buildbot.twistedmatrix.com/builders/winxp32-py2.5-select/builds/573) with just the test changes, and then I [rebuilt with the fix](http://buildbot.twistedmatrix.com/builders/winxp32-py2.5-select/builds/574).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [27580]) Branching to 'stdio-connection-lost-reason-3922'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [27594]) Merge stdio-connection-lost-reason-3922

Author: dotz, exarkun
Reviewer: therve
Fixes: [#3922](https://github.com/twisted/twisted/issues/3922)

Change Windows stdio support so that it passes a `Failure` to the protocol's
`connectionLost` method, rather than an exception.  Also adjust one of the
existing stdio tests to verify this behavior.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
